src/splice.test-d.ts

```diff
@@ -1,21 +1,179 @@
+    expectTypeOf(splice([] as Array<number>, 0, 0, [])).toEqualTypeOf<
+      Array<number>
+    >();
+  });
+
+  it("reflects the type of `replacement` in the return value", () => {
+    expectTypeOf(splice([], 0, 0, [] as Array<number>)).toEqualTypeOf<
+      Array<number>
+    >();
+  });
+
+  it("reflects the type of `replacement` in the return value (data-last)", () => {
+    expectTypeOf(pipe([], splice(0, 0, [] as Array<number>))).toEqualTypeOf<
+      Array<number>
+    >();
+  });
+
+  it("does not work with replacement of different type", () => {
+    // @ts-expect-error - wrong `replacement` type
+    splice([1, 2, 3] as const, 0, 0, ["a"] as const);
+  });
 });
 
-it("reflects the type of `replacement` in the return value", () => {
-  const replacement: Array<number> = [];
-  const result = splice([], 0, 0, replacement);
-  expectTypeOf(result).toEqualTypeOf<Array<number>>();
+describe("fixed tuples", () => {
+  describe("deleteCount = 0", () => {
+    it("works with start + length < 0", () => {
+      expectTypeOf(
+        splice([1, 2, 3] as const, -4, 0, [4] as const),
+      ).toEqualTypeOf<[4, 1, 2, 3]>();
+    });
+
+    it("works with start < 0", () => {
+      expectTypeOf(
+        splice([1, 2, 3] as const, -1, 0, [4] as const),
+      ).toEqualTypeOf<[1, 2, 4, 3]>();
+    });
+
+    it("works with start = 0", () => {
+      expectTypeOf(
+        splice([1, 2, 3] as const, 0, 0, [4] as const),
+      ).toEqualTypeOf<[4, 1, 2, 3]>();
+    });
+
+    it("works with start > 0", () => {
+      expectTypeOf(
+        splice([1, 2, 3] as const, 1, 0, [4] as const),
+      ).toEqualTypeOf<[1, 4, 2, 3]>();
+    });
+
+    it("works with start = length", () => {
+      expectTypeOf(
+        splice([1, 2, 3] as const, 3, 0, [4] as const),
+      ).toEqualTypeOf<[1, 2, 3, 4]>();
+    });
+
+    it("works with start > length", () => {
+      expectTypeOf(
+        splice([1, 2, 3] as const, 6, 0, [4] as const),
+      ).toEqualTypeOf<[1, 2, 3, 4]>();
+    });
+  });
+
+  describe("deleteCount != 0", () => {
+    it("works with deleteCount < 0", () => {
+      expectTypeOf(
+        splice([1, 2, 3] as const, 1, -1, [4] as const),
+      ).toEqualTypeOf<[1, 4, 2, 3]>();
+    });
+
+    it("works with deleteCount + start < length", () => {
+      expectTypeOf(
+        splice([1, 2, 3] as const, 1, 2, [4] as const),
+      ).toEqualTypeOf<[1, 4]>();
+    });
+
+    it("works with deleteCount + start > length", () => {
+      expectTypeOf(
+        splice([1, 2, 3] as const, 1, 4, [4] as const),
+      ).toEqualTypeOf<[1, 4]>();
+    });
+  });
+
+  it("works with empty replacement", () => {
+    expectTypeOf(splice([1, 2, 3] as const, 1, 1, [] as const)).toEqualTypeOf<
+      [1, 3]
+    >();
+  });
+
+  it("works with big replacement", () => {
+    expectTypeOf(
+      splice([1, 2, 3] as const, 1, 1, [4, 5, 6] as const),
+    ).toEqualTypeOf<[1, 4, 5, 6, 3]>();
+  });
+
+  describe("non-literal params", () => {
```

Comment:
The describe here isn't exact, 1 | 2 is considered a literal (at least by type-fest's defintion).